Respin how-to guide (#57)
canonical/kernel-docs: docs/how-to/respin.md

Next, on Jira, fill in the details of the respin in the description. This can include:

* Details on why the kernel is being respun
* Links to relevant bugs
* Affected kernels that will need to be respun

```{note}
Note the respin number, as it's used in subsequent steps.

The respin number is in the title of the Jira card in parentheses.
For example, "Re-spin (#8)" indicates the number is 8. 
```

## 2. Create tracking bugs
Each affected kernel needs a new Launchpad tracking bug.

Use the [kteam-tools/stable/create-kernel-tasks] script:

```text
create-kernel-tasks --handle <handle> --spin <spin-number> <cycle-number>
```

For example, we used:
```{terminal}
:input: create-kernel-tasks --handle noble:linux-gke --spin 8 2025.01.13
:dir: kteam-tools/stable/
```

The output states that a new tracking bug has been made, and also prints the bug number for the original tracking bug for this kernel's original crank.

On Launchpad, mark the original tracking bug as a duplicate of the newly-created bug:

<!-- TODO screenshots of how to do this -->

## 3. Checkout the last cranked version of the kernel

Use [cranky]() to checkout the kernel:

```text

> cranky checkout --cleanup --pristine <handle>

Reviewer: Is \<handle\> defined somewhere?

Author: I'm not sure if it's defined _formally_... 

Informally, it's `series:package`, where `package` is `linux` or `linux-<derivative>`

It may be formally defined in kteam-tools libs? I'll look around later. We can definitely write a little reference or something about it, too, so it's in the docs.

```

Then, use `cranky rmadison` to get the version number from the #proposed kernel:
```text
cranky rmadison <handle>
```

In `linux-main/`, ensure `HEAD` is on the tag of the kernel in #proposed (the output of `cranky rmadison`).

<!-- TODO what if it isn't (new commits were applied)? -->

<!-- For example, when running `git show` in `linux-main/`, we saw that `HEAD` was on `Ubuntu-` -->

## 4. Add patchset

The following step depends on whether the kernel you are respinning has a parent kernel which itself is a derivative.

## 5. Continue cranking, with some modifications

Run the following commands as normal:
```text
cranky open
cranky review-master-changes
```

In theory, no changes should be detected by this step.

Next, run `cranky link-tb` with a modified cycle number:
```text
cranky link-tb --sru-cycle <cycle-number>-<spin-number>
```

<!-- TODO example -->

Next,
```text
cranky update-dkms-versions
```
In theory, no changes should be detected by this step.

Next,
```text
cranky close 
cranky update-dependents

> cranky tags --force

Reviewer 2: any reason why `--force` is being used here? i think it shouldn't be necessary since the ABI has to be bumped anyway

Author: Good catch, I think we can get away with no `--force`

# How to respin a kernel

> Sometimes regressions or last-minute changes are identified in kernels in #proposed (already cranked).

Reviewer 3: We usually use `-proposed`.

> Sometimes regressions or last-minute changes are identified in kernels in #proposed (already cranked).

Reviewer 4: ```suggestion
Sometimes regressions or last-minute changes are identified in kernels in -proposed (already cranked).
```

> It is possible to make changes before the SRU cycle closes. This is a special version of a crank called a respin.

Reviewer 4: ```suggestion
It is possible to make changes before the SRU cycle closes. This special version of a crank is called a respin.
```

This makes "respin" more of a definition.

> These are the steps to respin a kernel, with examples from a previous `2025.01.13` respin of the `noble:linux-gke` kernel due to a late-cycle decision to revert a patchset.

Reviewer 4: Hey @benjamin051000 , i haven't finished reviewing this but a how-to shouldn't be too specific in terms of the scenarios described in the guide.
I would caution setting up this expectation by mentioning a specific cycle and kernel and reason for respin.

## 1. Create a respin Jira card

```{attention}
Only one person peforms this step. If you already have a Jira card, note the respin number in the card (see Note at the end of this step) and skip this step.

> ```

Reviewer 4: I would use Warning / Attention / Caution for something that you absolutely don't want someone to do without understanding the impact.

So in this case, if you DO create another Jira card, what happens?

Use [kteam-tools create-respin-card](https://kernel.ubuntu.com/forgejo/kernel/kteam-tools/src/branch/master/stable/create-respin-card) script:

```text
./create-respin-card <cycle-number>
```

> This will create a respin card on Jira. (For example, here's the one we created: [KSRU-15479 (Jira)](https://warthogs.atlassian.net/browse/KSRU-15479))

Reviewer 3: It will stick around for a while. The Jira dashboard for SRU cycles currently has cards going back to June 2024. It's probably best to add a screenshot.

> Use [kteam-tools create-respin-card](https://kernel.ubuntu.com/forgejo/kernel/kteam-tools/src/branch/master/stable/create-respin-card) script:
> 
> ```text
> ./create-respin-card <cycle-number>
> ```
> 
> This will create a respin card on Jira. (For example, here's the one we created: [KSRU-15479 (Jira)](https://warthogs.atlassian.net/browse/KSRU-15479))

Reviewer 4: ```suggestion
Create a respin card on Jira using the [kteam-tools create-respin-card](https://kernel.ubuntu.com/forgejo/kernel/kteam-tools/src/branch/master/stable/create-respin-card) script:

```text
./create-respin-card <cycle-number>
```

(For example, here's the one we created: [KSRU-15479 (Jira)](https://warthogs.atlassian.net/browse/KSRU-15479))
```

Reviewer 4: I suggest describing the step before showing the command
"Create a Jira card" instead of "Use kteam-tools" script which doesn't provide context.

> This will create a respin card on Jira. (For example, here's the one we created: [KSRU-15479 (Jira)](https://warthogs.atlassian.net/browse/KSRU-15479))

Author: I'll add it once this moves to the other repo

Next, on Jira, fill in the details of the respin in the description. This can include:

* Details on why the kernel is being respun
* Links to relevant bugs
* Affected kernels that will need to be respun

```{note}
Note the respin number, as it's used in subsequent steps.

The respin number is in the title of the Jira card in parentheses.
For example, "Re-spin (#8)" indicates the number is 8. 
```

## 2. Create tracking bugs

> Each affected kernel needs a new Launchpad tracking bug.

Reviewer 4: This is good. A bit of context should be provided as to why you need each step (where possible). Add something similar for step 1.

Use the [kteam-tools/stable/create-kernel-tasks] script:

```text
./create-kernel-tasks --handle <handle> --spin <spin-number> <cycle-number>
```

For example, we used:
```{terminal}
:input: ./create-kernel-tasks --handle noble:linux-gke --spin 8 2025.01.13

> :dir: kteam-tools/stable/

Reviewer: \:input\:
\:dir\:

These are showing up in the rendered doc.  Not sure that is the intention.

Author: They do on my machine 🤦🏼‍♂️

Author: I see them in https://canonical-kernel-docs--57.com.readthedocs.build/en/57/how-to/respin/, you don't? Maybe try a different browser?

```

The output states that a new tracking bug has been made, prints its bug number, and also prints the bug number for the original tracking bug from this kernel's original crank.

On Launchpad, mark the original tracking bug as a duplicate of the newly-created bug.

> <!-- TODO screenshots of how to do this -->

Author: It would be nice to have a screenshot of how to do this in launchpad, it's not obvious to newbies (myself).

## 3. Checkout the last cranked version of the kernel

Use cranky to checkout the kernel:

```text
cranky checkout --cleanup --pristine <handle>
```

Then, use `cranky rmadison` to get the version number from the #proposed kernel:
```text
cranky rmadison <handle>
```

> In `linux-main/`, ensure `HEAD` is on the tag of the kernel in #proposed (the output of `cranky rmadison`).

Author: Sometimes, there are multiple entries for package, see `bionic:linux-raspi-5.4`. Do you just choose the latest one in this case?

> <!-- TODO what if it isn't (new commits were applied)? -->

Author: What if new commits were applied? Does this ever happen?

Reviewer 2: yes, i think it does happen (someone please correct me if im wrong), but realistically only for first-level derivatives (e.g. derived directly from the generic kernels). for example, during a security cycle it's not uncommon for you to have commits that will have to wait for the next regular cycle to be included, so when you prepare the kernel for the security cycle you revert the head to the previous tag, crank it, and then reapply the commits. this leaves you with a couple commits on top of the latest tag in proposed

Reviewer 2: i think that in the case that the HEAD doesn't match the latest tag, you can either:

- work on the usual `cranky/master-next` branch, `git format-patch` the additional queued-up commits, `git revert --hard` back to the latest tag, crank as usual and then reapply the patches with `git am` on top of the respin tag
- create a new branch (e.g. `respin`), `git revert --hard` back to the latest tag, crank as usual, then go back to `cranky/master-next` and rebase it on top of the `respin` branch

<!-- For example, when running `git show` in `linux-main/`, we saw that `HEAD` was on `Ubuntu-` -->

## 4. Add patchset

```{attention}
This step only applies to kernels which don't have a parent with patches applied. If your kernel has a parent kernel where respin patches were applied, skip this step.
```

Usually patches will come from a mailing list, as `.patch` files. Use `git am` to apply them.

You may also be asked to modify the commit message.

> <!--TODO elaborate?-->

Author: Should we elaborate? Is there a standard/convention we generally follow?

## 5. Rebase
```{attention}
This step only applies to kernels which are derivatives of kernels also being respun. Skip this step if your kernel has a parent that isn't being respun.
```

Your kernel's parent will have a new tag from its respin. Rebase off that tag:
```text
cranky rebase -b <parent-respin-tag>
```

For `noble:linux-gke`, the parent kernel, `noble:linux`, was not respun, so we skipped this step.

## 6. Continue cranking, with some modifications

Run the following commands as normal:
```text
cranky open
cranky review-master-changes
```

> In theory, no changes should be detected by this step.

Author: Are there ever changes spotted here?

Reviewer 2: i think technically yes if the parent kernel had to be re-spun as well in order to change something inside its debian directory (which would mean someone would have to manually replicate that change in the derivative's debian directory when re-spinning). but it is rare i think

Next, run `cranky link-tb` with a modified cycle number:
```text
cranky link-tb --sru-cycle <cycle-number>-<spin-number>
```

In our crank, we used `cranky link-tb --sru-cycle 2025.01.13-8`, where 8 was the spin number.

Next, run:
```text
cranky update-dkms-versions
```

> In theory, no changes should be detected by this step.

Author: Are there ever changes spotted here?

Reviewer: I didn't see any for my respins.

Reviewer 2: i think technically yes if the parent kernel had to be re-spun as well in order to change something inside its `debian` directory (which would mean someone would have to manually replicate that change in the derivative's `debian` directory when re-spinning). but it *is* rare i think

Reviewer 3: Similar to the question above about review-master-changes, if a kernel was respun due to a necessary dkms change, changes could also appear here. Usually there would be no change though.

Next, finish the crank:
```text
cranky close 
cranky update-dependents
cranky tags
```

## 7. Pull sources
The primary difference with pulling the previous deb package sources is that we need to specify to pull the ones in #proposed. (By default, `cranky pull-sources` pulls TODO.)

First, we need to know which versions are in #proposed. Run `cranky rmadison` to get this info:
```text
cranky rmadison <handle>
```

You can `grep` the result for "proposed", too.

Note both the package name `cranky rmadison` produces and the version numbers of each package this kernel has (main, meta, signed, etc.) that are in #proposed.

For each package, we will run `cranky pull-source`.

```{attention}
Note that the command is `cranky pull-source`, _not_ `cranky pull-sources` (no 's').
We use `cranky pull-source` on each package manually so that we can specify the correct version number for each one.

> ```

Author: Does `pull-sources` have something like `build-source`'s `--build-opts` to pass the versions? That would be nice.

Reviewer 2: no, but you can use `cranky pull-source` (note the singular instead of plural) to specify the exact package version (but you have to do it for each of `main`,`meta`,`signed`,`lrm` separately). the reason why `pull-sources` doesn't have it is because the the other packages don't always follow the version from the main package (since there are frequent respins for them, especially `lrm`)

Reviewer 2: in my experience both `cranky pull-source` and `cranky pull-sources` fail to find the package im looking for so i frequently just pull the sources directly with `pull-ppa-source` (from the `ubuntu-dev-tools` package):

```bash
pull-ppa-source --ppa canonical-kernel-team/proposed2 linux-ibm noble
```

(note that you can also specify a package version instead of the series). the disadvantages being:

- you still have to do it for each package separately
- you have to know in which PPA the package you're looking for is (usually either `proposed` or `proposed2`, but you can check with `cranky list-routing`)

Then, use `cranky pull-source` with the following syntax:
```text
cd ..
cranky pull-source --no-verify <rmadison-package-name> <version-in-proposed> <series>
```

For example, the rmadison format looked like this (at time of our respin) for `noble:linux-gke`

```{terminal}
:input: cranky rmadison noble:linux-gke

> :dir: noble/linux-gke/

Reviewer: Same comment as above, showing up in rendered doc..

 linux-meta-gke   | 6.8.0-1017.21 | noble-proposed#2 | source 
 linux-signed-gke | 6.8.0-1003.5  | noble            | source 
 linux-signed-gke | 6.8.0-1017.21 | noble-security   | source 
 linux-signed-gke | 6.8.0-1017.21 | noble-updates    | source 
 linux-signed-gke | 6.8.0-1019.23 | noble-proposed   | source 
 linux-signed-gke | 6.8.0-1017.21 | noble-proposed#2 | source
```

`noble:linux-gke` has `main`, `meta`, and `signed` packages, so we noted those specific package names and versions from `cranky rmadison`:

| package name | version in `noble-updates` |
| - | - |
| `linux-gke` | `6.8.0-1017.21` |
| `linux-meta-gke` | `6.8.0-1017.21` |
| `linux-signed-gke` | `6.8.0-1017.21` |

```{note}
In the example, the version numbers for each package are the same.
This is not always the case, so it's still important to go through this step for each package.
```

So, we ran the following commands:
```{terminal}
:input: cranky pull-source linux-gke 6.8.0-1017.21 noble
:dir: noble/linux-gke/
```

```{terminal}
:input: cranky pull-source linux-meta-gke 6.8.0-1017.21 noble
:dir: noble/linux-gke/
```

```{terminal}
:input: cranky pull-source linux-signed-gke 6.8.0-1017.21 noble

> :dir: noble/linux-gke/

Reviewer: These, too.

```

## 8. Build sources

We need to pass the information about the packages in #proposed gathered in the previous step (`pull-source`) to build the sources this time. 

> <!--TODO why?-->

Author: Why is this necessary? It has no choice but to use the packages we downloaded... right?

Reviewer: You could have stale ones lying around if you didn't run a clean step prior to the respin.

Author: Ah, good point.

`cranky build-sources` has a `--build-opts` flag we can use to specify the version to use. Pass this flag once for each package, like so:

```bash
cd linux-main/
cranky build-sources --build-opts 'main:-v<version-in-updates>' --build-opts 'meta:-v<version-in-updates>' --build-opts 'signed:<version-in-updates>'
```

```{note}
The syntax used for each `--build-opts` is `<package-dirname>:-v<version>`.
The `package-dirname` is the name of the directory this package lives in (use `ls` to see them).
They are usually `linux-main/`, `linux-meta/`, `linux-lrm`, `linux-signed/`, etc.
Use these without the `linux-` prefix.
```

For example, we used the following command to build `noble:linux-gke`:

```{terminal}
:input: cranky build-sources --build-opts 'main:-v6.8.0-1017.21' --build-opts 'meta:-v6.8.0-1017.21' --build-opts 'signed:-v6.8.0-1017.21'

> :dir: noble/linux-gke/linux-main

Reviewer: These, too.

```

## 9. Review changes and upload

> At this point, refer to the normal crank process to finish the respin.

Reviewer 2: i would add that its good to double check the debian changelog in the closing commit message for the main package (to make sure that it looks the same as the previous one + the change you made) and for the other packages (to make sure that you included the changes from the pre-respin version)

```bash
cd ..
cranky review *.changes
```

> Lastly, upload the package or push it for a peer-review.

Reviewer 2: (if you dont have upload rights yet)
